### ISSUE TITLE
can't seem to get this to work

### ISSUE THREAD

**Issue by tobydeemer** [raburton/tabscroller#6]
apologies if I'm missing any steps, etc., but I can't get TabScroller to work. Right-click + scrolling doesn't do anything when mouse is over web page content. Only thing that happens is the right-click menu doesn't pop until I release the button.

I've removed and re-added the addon a few times to test. Is there anything in about:config I could check for conflicts? The only thing there I note with "scroll" in it that seems may be relevant are the entries from the addon itself, but I thought it would be worth a check.

Thanks though for the time and effort in making this addon.

**Comment by tobydeemer** [comment]
Err, sorry-

I am FF 60.0.1 (64-bit) on Mac OS.

Tab Scroller is 1.0.5.

Thanks.

**Comment by raburton** [comment]
Hi, thanks for reporting the issue. MacOS is the only platform I can't test on, but I'm assured that it was working by people who have tried it, though not specifically had feedback about 60.0.1
Did you have it working previously?
Do you have any other add-ons installed? I can try those here on Windows and Linux to see if there is some form of conflict.
Sounds like the right click menu has been changed to mouse up, that's necessary so it's good that that's working. If you right click and scroll (hoping for the tab change, but not getting it) does that prevent the menu popping up when you release the button?
And just to confirm some basics, just to be sure - you say right click, but you are continuing to hold the button down while scrolling? You're testing on a few normal web pages, not pages from the mozzila add-ons site or special pages like the new tab page or firefox options page?
Might be I have to give you a debug version to make sure the messages are being passed correctly, and the webextension api is getting called.

**Comment by tobydeemer** [comment]
> Did you have it working previously?

nope.

> Do you have any other add-ons installed? I can try those here on Windows and Linux to see if there is some form of conflict.

yep:
CheckThemAll
Dark Background and Light Text
HTTPS Everywhere
Privacy Badger
TabScroller (obvs...)
uBlock Origin
uMatrix

> Sounds like the right click menu has been changed to mouse up, that's necessary so it's good that that's working. If you right click and scroll (hoping for the tab change, but not getting it) does that prevent the menu popping up when you release the button?

No, when I hold right click and scroll, nothing happens typically but on pages with scrollable content, it just scrolls that. And then when I release the mouse button, the context menu pops.

> And just to confirm some basics, just to be sure - 
> you say right click, but you are continuing to hold the button down while scrolling? 

yes

> You're testing on a few normal web pages, not pages from the mozzila add-ons site or special pages like the new tab page or firefox options page?

yep.


Thanks so much for the time.

**Comment by tobydeemer** [comment]
update: 

I just tried it with disabling (not removing) all other addons, and no change.

**Comment by raburton** [comment]
Ok, made a debug version you can try:
https://www.dropbox.com/s/z28771hfmmav6ph/tabscroller-debug.zip?dl=0
Extract the zip file somewhere on your mac.
Open a new firefox tab and go to this address: about:debugging#addons
Enable add-on debugging. Click load temporary addon and navigate to the extracted folder from above and select the manifest file. This should load the debug version of tab scroller.
Click the debug link under the addon. This will open a windows for the background thread console.
Click the firefox menu drop down and go to web developer submenu and select brower console (not web console). This will open another similar looking window.
Now in a new tab, opened since loading the debug plugin, open a couple of normal web pages and try to use the addon on them.

In the addon debug windows I gets lots of messages like this as I scroll back and forward:
```
msg scrollDown
msg scrollUp
msg scrollDown
msg scrollUp
msg scrollDown
msg scrollUp
msg scrollDown
```
In the browser console window I get messages like this:
```
mousedown: button 2
wheel: buttons 2
scrollDown
msg scrollDown
wheel: buttons 2
scrollUp
msg scrollUp
wheel: buttons 2
scrollDown
msg scrollDown
TypeError: selection is null contextmenuhlpr.js:8:1
wheel: buttons 2
scrollUp
msg scrollUp
contextmenu: block  true
```

**Comment by tobydeemer** [comment]
tried a couple times:

https://i.imgur.com/GuxJKDp.png

https://i.imgur.com/IbXdGcO.png

**Comment by raburton** [comment]
Not a single entry there at all from the addon. Assuming the temporary one overrode the original (looks like it from the message about the old version being disabled, but just double check the main addons page, click more next to tabscroller and ensure it says 1.0.5a, not 1.0.5) then it doesn't look like it's really being loaded/executed at all. One way to check for sure, on a normal page right click and choose inspect element, on the new section that opens at the bottom of the page view the debugger tab and expand moz-extension and see if the content.js script is loaded liek this: https://imgur.com/a/vbWrfSS
You can also look at the console tab, click the bin icon to clear it, then refresh the page and see if anything odd comes up in the console.
I note a some messages saying the security policy blocked some scripts being loaded at resource "self", which is a bit suspicious for something internal (like an addon) being blocked, and not an error I see here, but could be a red herring.

**Comment by tobydeemer** [comment]
Well, may be worth looking at version issues. I removed it, restarted FF, re-added, and it's still 1.0.5., and 1.0.5a isn't showing as available on the add-on site. Is there an updated version available directly?

**Comment by raburton** [comment]
The new version only contains a lot of debug, so I don't want to post that
to the addons site or everyone will end up getting it. I've checked the zip
file I sent you and the manifest definitely says 1.0.5a so it doesn't look
like it's loaded properly. Best thing to do it remove tabscroller
altogether, then just do the load of the temporary addon from the contents
of the zip file I sent, that way there will be no conflict, you'll only
have the one version on the system.

**Comment by tobydeemer** [comment]
Ah, ok. Makes sense. I'll try that and report back again.

**Comment by tobydeemer** [comment]
ok-

I'm getting no messages in the add-on debug window, and in the browser console window I get this:

https://i.imgur.com/GBb6po1.png

and clicking on one of the entries for "content.js" with the red counter next to it, it opens a file:/// link:

view-source:moz-extension://d62be84c-6706-aa42-be0b-88d2fc28ca46/content.js

which contains:

'use strict';

const RIGHT_BUTTON = 2;

// flag to indicate we should block the context menu
let blockContextMenu = false;

// message from background to indicate tab has been scrolled
// to and that we should disable the context menu for when
// the user releases the mouse button
browser.runtime.onMessage.addListener((message, sender) => {
	switch (message.topic) {
	case 'scrolledToTab':
		blockContextMenu = true;
		break;
	}
	return false;
});

window.addEventListener('mousedown', function (event) {
	// re-enable context menu on right button press
	// (will be disabled again if users scrolls while holding)
	console.log("mousedown: button %d", event.button);
	if (blockContextMenu && (event.button === RIGHT_BUTTON)) {
		blockContextMenu = false;
	}
}, true);

window.addEventListener('contextmenu', function (event) {
	// prevent context menu if we've been scrolling
	console.log("contextmenu: block %o", blockContextMenu);
	if (blockContextMenu) {
		event.preventDefault();
		event.stopPropagation();
	}
}, true);

window.addEventListener('wheel', function (event) {

	console.log("wheel: buttons %d", event.buttons);

	// only act if right mouse button held down
	if (event.isTrusted && (event.buttons & RIGHT_BUTTON)) {

		// prevent page scrolling
		event.preventDefault();
		event.stopPropagation();

		// prevent context menu when button released
		blockContextMenu = true;

		// send tab scroll request to background
		if (event.deltaY > 0) {
			console.log("scrollDown");
			browser.runtime.sendMessage({
				topic: 'scrollDown'
			});
		} else if (event.deltaY < 0) {
			console.log("scrollUp");
			browser.runtime.sendMessage({
				topic: 'scrollUp'
			});
		}
	}

}, true);

**Comment by tobydeemer** [comment]
[browser_console_full_output.txt](https://github.com/raburton/tabscroller/files/2035334/browser_console_full_output.txt)

attached is a txt file of the full browser console output.

I tried it with uBlock enabled, then disabled. Leaving uBlock disabled, I then tried it with uMatrix disabled, and last attempt was with uMatrix only re-enabled. The results were the same on each attempt, but console output seems to be slightly different.

**Comment by raburton** [comment]
Sorry for the slow reply, I had to week a long weekend on-call, then took a couple of days back to recover. So, what I see from the log is that when the mouse scroll wheel event is called it does not show that any buttons where held down at the time.
According to Mozilla docs this is to be expected with OSX 10.5, it doesn't say "10.5 or later", but perhaps that's what it means. Going against that is that I've had a couple of people tell me it works, though I don't know what version of OSX they were using. I'll try and clarify if it should work on newer versions.

**Comment by tobydeemer** [comment]
No worries at all. I know how on-call goes. Thanks again for the efforts.

**Comment by raburton** [comment]
I haven't given up on this - would you try this and see if it works for you: https://www.dropbox.com/s/3az2a19vq2oh9hq/tabscroller-mac.zip?dl=1

**Comment by tobydeemer** [comment]
Success!! I loaded the manifest, and it just worked. I assume this needs to be pushed out in order to be added via regular extensions? 

Thanks so much again for your time; this is awesome.

**Comment by tobydeemer** [comment]
Weird behavior though-  if I manually click a tab it activates scrolling-to-switch-tabs. Have to right-click to "release" the right button, to disable tab switching.

**Comment by raburton** [comment]
Thanks for the feedback, I'll look into this issue...

**Comment by tobydeemer** [comment]
No problem at all, and thank you again for all your work on this. I know
dev time like this isn't easy, and this function greatly helps my work
flow. It's very much appreciated.

---
This message was decidedly not sent from an iPhone.

**Comment by raburton** [comment]
Fixed. I'm away for a week now, if you give it a test and don't find any other problems I'll upload it when I get back.
https://www.dropbox.com/s/uflieyt15edznrr/tabscroller-mac.zip?dl=1

**Comment by raburton** [comment]
Any joy?

**Comment by tobydeemer** [comment]
Yes, sorry for the late reply, but it looks great. Thanks again!

**Comment by raburton** [comment]
Closed in v1.0.6
